### PR TITLE
Lancaster stemmer AE

### PR DESCRIPTION
**Pull request by mjunsilo**
This Lancaster stemmer is based on the on the Smile machine learning package (see http://haifengl.github.io/smile). This AE implementation was derived from the SnowballStemmer codebase.

### CODE REVIEW

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
+package de.tudarmstadt.ukp.dkpro.core.lancaster;
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: License header is missing.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import org.apache.uima.jcas.JCas;
+
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+
+/**
+ * This Paice/Husk Lancaster stemmer implementation only works with the English language so far.
+ *
+ * TODO: The wrapped Lancaster stemmer of the Smile API should be enhanced to support custom rules files.
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: Todo should be outside the JavaDoc comment as the JavaDoc comment is used to generate the reference documentation.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        extends FeaturePathAnnotatorBase {
+
+    private static final String MESSAGE_DIGEST = LancasterStemmer.class.getName()+"_Messages";
+
+    /**
+     * True if the stemmer will strip prefix such as kilo, micro, milli, intra, ultra, mega, nano, pico, pseudo.
+     */
+    public static final String PARAM_STRIP_PREFIXES = "stripPrefix";
+    @ConfigurationParameter(name = PARAM_STRIP_PREFIXES, mandatory = false, defaultValue="false")
+    protected boolean stripPrefix;
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: All fields and methods should by default be private. Normally, components do not get subclassed.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            StemAnnotator stemAnnotator = new StemAnnotator(jcas);
+
+            for (String path : paths) {
+
+                // Separate Typename and featurepath
+                String[] segments = path.split("/", 2);
+                String typeName = segments[0];
+
+                // Try to get the type from the typesystem of the CAS
+                Type t = currCAS.getTypeSystem().getType(typeName);
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: Use CasUtil.getType(). It also created the "type not found" exception.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                            // check annotation filter condition
+                            if (this.filterFeaturePathInfo.match(fs, this.filterCondition)) {
+                                stemAnnotator.createStemAnnotation(fs);
+                            }
+                        }
+                        else { // no annotation filter specified
+                            stemAnnotator.createStemAnnotation(fs);
+                        }
+                    }
+                    catch (AnalysisEngineProcessException e) {
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: Catching AnalysisEngineProcessException despite AnalysisEngineProcessException being declared in throws. This catch can be removed.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                        throw new IllegalStateException(
+                                "error occured while creating a stem annotation", e);
+                    }
+                }
+            }
+
+    }
+
+
+    protected class StemAnnotator {
```
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: I think there is no need for the inner class here. Simply move createStemAnnotation to the outer class.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java (hunk 1 above)
Comment (mjunsilo) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: SnowballStemmer also lacks a license header. Should I add this?

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java (hunk 3 above)
Comment (mjunsilo) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: Just saw this somewhere so I thought it was common practice. Anyway, its fixed now.

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java (hunk 1 above)
Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: https://github.com/dkpro/dkpro-core/blob/master/dkpro-core-snowball-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/snowball/SnowballStemmer.java has a header. Are you referring to a different SnowballStemmer?

File: dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java (hunk 6 above)
Comment (mjunsilo) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: Seems like a waste to initialise a new stemmer on every token but sure I can also pass it to the function on every call together with the JCas, although I then personally prefer some static or inner class depending on how many outer class member are accessed.

Comment (reckart) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: I would move the `stemmer` field to the outer class and initialize it once in the initialize() method. It is not a blocker. Can be refactored after merge.

Comment (mjunsilo) [inline dkpro-core-lancaster-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/lancaster/LancasterStemmer.java]: OK, I agree the code its a bit more compact now as a method. Fixed.

**Comment by mjunsilo** [comment]
I think I should have addressed all comments. Please let me know if anything still needs corrections.

**Comment by reckart** [comment]
Since this is a new module, it should use the groupId "org.dkpro.core" and the artifactId "dkpro-core-lancasterstemmer-asl". We are also a more community-oriented license header for new modules (cf. e.g. https://github.com/dkpro/dkpro-core/blob/master/dkpro-core-io-lxf-asl/src/main/java/org/dkpro/core/io/lxf/LxfReader.java). You can update the PR once more, otherwise I 'll do it after the merge.

**Comment by reckart** [comment]
Oh, and you might want to add yourself to the CONTRIBUTORS.txt file in the project root ;)

**Comment by mjunsilo** [comment]
OK, I'll have a look at it tomorrow.

**Comment by mjunsilo** [comment]
Made the module name and license header changes now.

**Comment by reckart** [comment]
Thanks!

**Pull request merged**